parastell/utils.py

```diff
@@ -227,32 +285,101 @@ def reorder_loop(list, index):
-                sigma=sigma,
-                mode="wrap",
-            ),
+    # Define small value
+    eps = 1e-10
+    # Shift some coordinates by eps to compute appropriate midplane flags
+    # If z = 0, then midplane flag = 0, then incorrect maximum radius computed,
+    # then incorrect OB midplane index
+    # Replace z-coordinates at 0 with eps
+    np.place(coords[:, 1], np.abs(coords[:, 1]) < eps, [eps])
+
+    radii = coords[:, 0]
+    # Determine whether adjacent points cross the midplane (if so, they will
+    # have opposite signs)
+    shifted_coords = np.append(coords[1:], [coords[1]], axis=0)
+    midplane_flags = -np.sign(coords[:, 1] * shifted_coords[:, 1])
+
+    # Count number of crossing points (will have positive magnitude)
+    count = np.count_nonzero(midplane_flags > 0)
+
+    # If no crossing points, loop does not cross midplane
+    if count == 0:
+        # Find index of point closest to midplane
+        obmp_index = np.argmin(np.abs(coords[:, 1]))
+    # If 1 crossing point, coordinates likely do not represent a closed loop
+    elif count == 1:
+        e = AssertionError(
+            "Only one crossing point found. Coordinates given likely do not "
+            "represent a closed loop."
+        )
+        raise e
+    # If 2 or more crossing points, loop crosses midplane
+    elif count >= 2:
+        # Find index of outboard midplane point
+        obmp_index = np.argmax(midplane_flags * radii)
+
+        # Handle edge cases
+        if obmp_index == 0:
+            previous_idx = -2
+        else:
+            previous_idx = obmp_index - 1
+
+        if obmp_index == len(coords) - 1:
+            next_idx = 1
+        else:
+            next_idx = obmp_index + 1
+
+        # Check adjacent points
+        obmp_index = obmp_index + (
+            np.argmin(
+                np.abs(
+                    [
+                        coords[previous_idx, 1],
+                        coords[obmp_index, 1],
+                        coords[next_idx, 1],
+                    ]
+                )
+            )
+            - 1
         )
-        previous_matrix = smoothed_matrix
 
-    return smoothed_matrix
+        # Handle edge cases
+        if obmp_index == len(coords) - 1 or obmp_index == len(coords):
+            obmp_index = 0
+
+    return obmp_index
+
+
+def orient_coords(coords, positive=True):
+    """Orients closed loop coordinates such that they initially
+    progress positively or negatively in the z-direction.
+
+    Arguments:
+        coords (Nx2 numpy.array of float): list of R, Z coordinates for the
+            closed loop.
+        positive (bool): progress coordinates in positive z-direciton
+            (defaults to True). If negative, coordinates will progress in
+            negative direction.
+
+    Returns:
+        coords (Nx2 numpy.array of float): reordered list of R, Z coordinates
+            for the closed loop.
+    """
+    if positive == (coords[0, -1] > coords[1, -1]):
```

Comment:
Does this assume that the first entry is on the outboard?  I am not sure I see all the places this is used, but if the first entry is on the inboard, then we want `positive = False`.  Not sure how we test that?